core/src/main/scala/org/apache/spark/ContextCleaner.scala

```diff
@@ -58,7 +58,12 @@ private class CleanupTaskWeakReference(
  */
 private[spark] class ContextCleaner(sc: SparkContext) extends Logging {
 
-  private val referenceBuffer = new ConcurrentLinkedQueue[CleanupTaskWeakReference]()
+  /**
+   * A buffer to ensure that `CleanupTaskWeakReference`s are not garbage collected as long as they
+   * have not been handled by the reference queue.
+   */
+  private val referenceBuffer =
+    new ConcurrentHashMap[CleanupTaskWeakReference, CleanupTaskWeakReference]()
```

Comment:
Hm, I wonder why? on its face it sounds less efficient to use a Map as a Set. If there's a good reason, then at least Collections.newSetFromMap(map) does sound better, I agree with that.